File: Data Structures/Data-structures-1/Queue/Circular-queue/circular-queue.py
#circular queue implementation in python:

class CircularQueue():
    def __init__(self, size):
        self.size = size
        self.queue = [None] * size
        self.queueReset()

    def isFull(self):
        return ((self.rear + 1) % self.size == self.front)
    def isEmpty(self):
        return(self.front == -1) 
    def justEmpty(self):
        return(self.rear == self.front)
    def queueReset(self):
        self.front = self.rear = -1
    def enqueue(self, data):
        if(self.isFull()):
            print("The queue is already full!")
        elif(self.isEmpty()):
            print("The queue is empty")
            self.front = self.rear = 0
            self.queue[self.rear] = data
        else:
            self.rear = (self.rear + 1) % self.size
            self.queue[self.rear] = data
    def dequeue(self):
        if(self.isEmpty()):
            print("The queue is already empty")
        if(self.justEmpty()):
            dequeued_value = self.queue[self.front]
            self.queueReset()
            return dequeued_value
        else:
            dequeued_value = self.queue[self.front]
            self.front = (self.front + 1) % self.size
            return dequeued_value
    def printCircularQueue(self):
        if (self.isEmpty()):
            print("The queue is empty!")
        elif(self.rear >= self.front):
            for i in range(self.front, self.rear + 1):
                print(self.queue[i], end=" ")
            print()
        else:
            for i in range(self.front, self.size):
                print(self.queue[i], end=" ")
            for i in range(0, self.rear + 1):
                print(self.queue[i], end=" ")
            print()

if __name__ == "__main__":
    cQueue = CircularQueue(5)
    for i in range(10):
        cQueue.enqueue(i)
    cQueue.printCircularQueue()
    for i in range(2):
        cQueue.dequeue()
    cQueue.printCircularQueue()

File: Data Structures/Data-structures-1/Queue/Circular-queue/Circular-queue.cpp

/**
 * circular queue implementation in cpp:
*/
#include "iostream"
#include "cstdio"

using namespace std;

class CircularQueue{
    private:
        //define the maximum queue size:
        static const int size = 5;
        int items[size], front, rear;
    public:
        CircularQueue(){
            front = -1;
            rear = -1;
        }
        bool isFull(){
            if(front == 0 && rear == size - 1){
                return true;
            }
            if (front == rear + 1){
                return true;
            }
            return false;
        }

        bool isEmpty(){
            if (front == -1){
                return true;
            }
            return false;
        }

        void enqueue(int element){
            if (isFull()){
                cout << "The circular queue is full!" << endl;
                return;
            } else{
                if (front == -1){
                    front = 0;
                }
                rear = (rear + 1) % size;
                items[rear] = element;
                cout << "Inserted " << element << endl;
                return;
            }
        }

        int dequeue(){
            int element;
            if(isEmpty()){
                cout << "The queue is empty" << endl;
                return -1;
            } else{
                element = items[front];
                /* queue has only one element therefore reset the queue*/
                if (front == rear){
                    front, rear = -1;
                } else{
                    front = (front + 1) % size;
                }
                return element;

            }
        }

        void printQueue(){
            int i;
            if(isEmpty()){
                cout << "Queue is empty!" << endl;
                return;
            } else{
                cout << "Front ->" << front << endl << "Items ->";
                for (i = front; i != rear; i = (i + 1)% size){
                    cout << items[i] << " ";
                }
                cout << items[i] << endl << "Rear -> " << rear << endl;
                return;
            }
        }
};

int main(){
    CircularQueue q;
    /* should fail as queue is empty*/
    q.dequeue();
    q.enqueue(1);
    q.enqueue(2);
    q.enqueue(3);
    q.enqueue(4);
    q.enqueue(5);
    /* should fail as queue is full*/
    q.enqueue(6);
    q.printQueue();

    if(int element = q.dequeue() != -1){
        cout << " Deleted -> " << element << endl;
    }
    q.printQueue();
    q.enqueue(7);
    q.printQueue();

    /*should fail*/
    q.enqueue(8);
}